### PR TITLE
tag_invoke based custom types

### PR DESCRIPTION
**Pull request by the-moisrex**
Now you can use tag_invoke to add a custom type or a group of custom types.

A follow up on [Daniel Lemire's tweet](https://x.com/lemire/status/1815544018979504145), here's a proposal on how we would use [tag_invoke](https://wg21.link/P1895) in simdjson.

There might be some small details like which namespace to use for some utilities, coding style, and other things that I have left after knowing the project's authors opinions and whether or not they want to go in this direction or not.

This allows you to:

```c++
struct Car {
  std::string make;
  std::string model;

  friend simdjson_result<Car>
  tag_invoke(simdjson::deserialize_tag, std::type_identity<Car>, simdjson::ondemand::value &val) {
    // ...
    return car;
  }
};
```

or completely going general, like:
```c++
namespace simdjson {

// this is to demonstrate that we can use concept; otherwise a simple
// type_identity<unique_ptr<T>> as the second argument would have done the job.
//
// This tag_invoke MUST be inside simdjson namespace
template <typename T>
  requires is_unique_ptr_v<T>
auto tag_invoke(deserialize_tag, std::type_identity<T>, ondemand::value &val) {
  using type = typename T::element_type;
  return simdjson_result{std::make_unique<type>(val.template get<type>())};
}

} // namespace simdjson
```

The use case would be:

```c++
  for (auto val : doc) {
    std::unique_ptr<Car> c(val); // <---- magically works
  }
```

### CODE REVIEW

**Comment by the-moisrex** [comment]
Also, even though I wrote it in C++20 for now until I know the authors opinions, there's nothing stopping us from re-implementing `tag_invoke` in C++17 without concepts.

**Comment by lemire** [comment]
@the-moisrex Speaking for myself, I find this very exciting.

File: include/simdjson/generic/ondemand/value.h [hunk 1]
```diff
@@ -7,10 +7,78 @@
 #include "simdjson/generic/ondemand/value_iterator.h"
 #endif // SIMDJSON_CONDITIONAL_INCLUDE
 
+#if __cplusplus >= 201703
```
Comment (lemire) [inline include/simdjson/generic/ondemand/value.h:10]: The feature macro test is [__cpp_concepts](https://en.cppreference.com/w/cpp/feature_test#cpp_concepts).

So maybe 

```cpp
#ifdef __cpp_concepts

#endif
```

Comment (the-moisrex) [reply include/simdjson/generic/ondemand/value.h:10]: Oh yeah; also it seems like I have mistakenly used `>=` instead of `>` here.

**Review by FranciscoThiesen — APPROVED** [review]
Awesome work @the-moisrex!

**Comment by the-moisrex** [comment]
There a problem with `noexcept`; this needs to be `noexcept`, but potentially it can throw. This is a API problem of the old way of custom types.

```c++
template <>
simdjson::simdjson_result<std::vector<double>>
simdjson::ondemand::value::get() noexcept {
  ondemand::array array;
  auto error = get_array().get(array);
  if (error) {
    return error;
  }
  std::vector<double> vec;
  for (auto v : array) {
    double val;
    error = v.get_double().get(val);
    if (error) {
      return error;
    }
    vec.push_back(val);
  }
  return vec;
}
```


How we should address this? Should I force noexcept on user's `tag_invoke`s too, drop the current `get` member function's noexcept, or add 2 `get` member functions with different noexcept specifiers?

**Comment by the-moisrex** [comment]
Update:
1. Used `__cpp_concepts`
2. Regarding `noexcept`, made it so that the old way of customizing will keep requiring `noexcept`, but the `tag_invoke` way will obey whatever the user chooses.

File: include/simdjson/generic/ondemand/value.h [hunk 2]
```diff
@@ -7,10 +7,79 @@
 #include "simdjson/generic/ondemand/value_iterator.h"
 #endif // SIMDJSON_CONDITIONAL_INCLUDE
 
+#include <version>
```
Comment (lemire) [inline include/simdjson/generic/ondemand/value.h:10]:
```suggestion
#if SIMDJSON_CPLUSPLUS20
#include <version>
#endif
```

But this requires the following PR:

https://github.com/simdjson/simdjson/pull/2220

**Comment by lemire** [comment]
The tests failure are probably all related to the missing `<version>` header which is required only when C++20 is present.

**Comment by the-moisrex** [comment]
I think I fixed `<version>` problem, but I'm not sure if that was the problem for the MSVC pipelines.

**Comment by lemire** [comment]
> I think I fixed <version> problem, but I'm not sure if that was the problem for the MSVC pipelines.

Your approach should work well.

Let us see what the tests say.

**Comment by the-moisrex** [comment]
Seems like my implementation of `tag_invoke` doesn't work on MSVC. I found a very weird implementation, let me debug this!

**Comment by lemire** [comment]
We getting...

```
D:\a\simdjson\simdjson\tests\ondemand\ondemand_custom_types_tests.cpp
  ondemand_custom_types_tests.cpp
D:\a\simdjson\simdjson\include\simdjson\generic\ondemand\value.h(123,19): error C2338: static_assert failed: 'The get method with given type is not implemented by the simdjson library. The supported types are ondemand::object, ondemand::array, raw_json_string, std::string_view, uint64_t, int64_t, double, and bool. We recommend you use get_double(), get_bool(), get_uint64(), get_int64(),  get_object(), get_array(), get_raw_json_string(), or get_string() instead of the get template. You may also add support for custom types, see our documentation.' [D:\a\simdjson\simdjson\build\tests\ondemand\ondemand_custom_types_tests.vcxproj]
```

This does not seem like a very troubling error.

**Comment by the-moisrex** [comment]
Jesus Christ!
Wasted so much time checking out different implementations of `tag_invoke`.

Turns out this `= delete` was the problem. I have no idea why! but it works now.

```c++
  void tag_invoke() = delete;
```

**Comment by the-moisrex** [comment]
Jesus Christ!
Wasted so much time checking out different implementations of `tag_invoke`.

Turns out this `= delete` was the problem. I have no idea why! but it works now.

```c++
  void tag_invoke() = delete;
```

**Comment by lemire** [comment]
@the-moisrex It is all green. Congratulation. We will merge this.

**Comment by FranciscoThiesen** [comment]
@the-moisrex do you know if this tag_invoke approach works for specializations that depend upon each other?

Let's say I first want to define a ondemand::value::get() for std::vector< int >. And then later I want to define an ondemand::value::get() for a struct X, that has std::vector<int> as one of their fields...

**Comment by the-moisrex** [comment]
> @the-moisrex do you know if this tag_invoke approach works for specializations that depend upon each other?
> 
> Let's say I first want to define a ondemand::value::get() for std::vector< int >. And then later I want to define an ondemand::value::get() for a struct X, that has std::vector as one of their fields...

I believe this is exactly what I've done in the example file `tests/ondemand/ondemand_custom_types_tests.cpp`.

First I defined customization for `vector<double>` here (with the old way of doing it just to show that we can combine these two ways):

```cpp
template <>
simdjson::simdjson_result<std::vector<double>>
simdjson::ondemand::value::get() noexcept {
  // ...
}
```

And here I used `vector<double>` in `Car`:

```cpp
struct Car {
  std::string make;
  std::string model;
  int64_t year = 0;
  std::vector<double> tire_pressure;  // <------ here it is

  // ...
};
```

And in `Car`'s `tag_invoke` we can use it like this or any other way:

```cpp
  friend simdjson_result<Car>
  tag_invoke(simdjson::deserialize_tag, std::type_identity<Car>, simdjson::ondemand::value &val) {
    for (auto field : obj) {
      // ...
      if (key == "tire_pressure") {
        error = field.value().get<std::vector<double>>().get(car.tire_pressure);
        // ...
      }
    }
    // ...
  }
```


------

And also the `unique_ptr` example shows this as well if you look at the implementation:
```cpp

namespace simdjson {

template <typename T>
  requires is_unique_ptr_v<T>
auto tag_invoke(deserialize_tag, std::type_identity<T>, ondemand::value &val) {
  using type = typename T::element_type;
  return simdjson_result{std::make_unique<type>(
        val.template get<type>() // <--------------- here we're getting the value of another type
  )};
}

} // namespace simdjson
```

That means during the call to this `tag_invoke` we might be calling another `tag_invoke` as well.

**Comment by lemire** [comment]
Let us merge this into a the simdjson repo, but within a dedicated branch. We will not be releasing this right now, because it needs some additional testing, documentation and maturation.

However, @the-moisrex : my expectation is that this will be part of a future release.

**Comment by lemire** [comment]
BTW this is really excellent work. My previous comment should not be interpreted as me being critical.

**Comment by lemire** [comment]
@FranciscoThiesen Merged to simdjson:builder_development_branch.

**Pull request merged**